File: workers/ruby/lib/tasker_core/types/step_context.rb
# frozen_string_literal: true

module TaskerCore
  module Types
    # StepContext provides a unified context for step handler execution.
    #
    # This is the cross-language standard context object passed to handler.call(context).
    # It wraps the FFI-provided TaskSequenceStepWrapper and adds convenience accessors
    # that match Python and Rust naming conventions.
    #
    # Cross-language standard fields:
    # - task_uuid: UUID of the task
    # - step_uuid: UUID of the workflow step
    # - input_data: Step input data from workflow_step.inputs
    # - step_inputs: Alias for input_data
    # - step_config: Handler configuration from step_definition.handler.initialization
    # - retry_count: Current retry attempt count
    # - max_retries: Maximum retry attempts allowed
    # - dependency_results: Results from parent steps
    #
    # Ruby-specific accessors (for backward compatibility):
    # - task: TaskWrapper instance
    # - workflow_step: WorkflowStepWrapper instance
    # - step_definition: StepDefinitionWrapper instance
    #
    # @example Accessing context in a handler
    #   def call(context)
    #     # Cross-language standard fields
    #     task_uuid = context.task_uuid
    #     step_uuid = context.step_uuid
    #     input_data = context.input_data
    #     deps = context.dependency_results
    #
    #     # Convenience methods
    #     even_number = context.get_task_field('even_number')
    #     prev_result = context.get_dependency_result('step_1')
    #
    #     # Ruby-specific accessors (backward compat)
    #     task = context.task
    #     step = context.workflow_step
    #   end
    #
    # @see TaskerCore::Models::TaskSequenceStepWrapper The underlying wrapper
    class StepContext
      # @return [TaskerCore::Models::TaskWrapper] Task metadata and context
      attr_reader :task

      # @return [TaskerCore::Models::WorkflowStepWrapper] Step execution state
      attr_reader :workflow_step

      # @return [TaskerCore::Models::DependencyResultsWrapper] Results from parent steps
      attr_reader :dependency_results

      # @return [TaskerCore::Models::StepDefinitionWrapper] Step definition from template
      attr_reader :step_definition

      # @return [String] The handler name for this step
      attr_reader :handler_name

      # Creates a StepContext from FFI step data
      #
      # @param step_data [Hash, TaskerCore::Models::TaskSequenceStepWrapper] The step data from Rust FFI
      # @param handler_name [String, nil] Optional handler name override
      def initialize(step_data, handler_name: nil)
        if step_data.is_a?(TaskerCore::Models::TaskSequenceStepWrapper)
          @task = step_data.task
          @workflow_step = step_data.workflow_step
          @dependency_results = step_data.dependency_results
          @step_definition = step_data.step_definition
        else
          wrapper = TaskerCore::Models::TaskSequenceStepWrapper.new(step_data)
          @task = wrapper.task
          @workflow_step = wrapper.workflow_step
          @dependency_results = wrapper.dependency_results
          @step_definition = wrapper.step_definition
        end

        @handler_name = handler_name || @step_definition.handler&.callable
      end

      # ========================================================================
      # CROSS-LANGUAGE STANDARD FIELDS
      # ========================================================================

      # Cross-language standard: task_uuid
      # @return [String] UUID of the task
      def task_uuid
        @task.task_uuid
      end

      # Cross-language standard: step_uuid
      # @return [String] UUID of the workflow step
      def step_uuid
        @workflow_step.workflow_step_uuid
      end

      # Cross-language standard: input_data
      # Returns the step inputs from the workflow step.
      # @return [ActiveSupport::HashWithIndifferentAccess] Step input data
      def input_data
        @workflow_step.inputs
      end

      # Cross-language standard: step_inputs (alias for input_data)
      # @return [ActiveSupport::HashWithIndifferentAccess] Step input data
      alias step_inputs input_data

      # Cross-language standard: step_config
      # Returns the handler configuration from the step definition.
      # @return [ActiveSupport::HashWithIndifferentAccess] Handler configuration from template
      def step_config
        @step_definition.handler&.initialization || {}.with_indifferent_access
      end

      # Cross-language standard: retry_count
      # @return [Integer] Current retry attempt count
      def retry_count
        @workflow_step.attempts || 0
      end

      # Cross-language standard: max_retries
      # @return [Integer] Maximum retry attempts allowed
      def max_retries
        @workflow_step.max_attempts || 3
      end

      # ========================================================================
      # CONVENIENCE METHODS
      # ========================================================================

      # Get a field from the task context.
      #
      # @param field_name [String, Symbol] Field name in task context
      # @return [Object, nil] The field value or nil if not found
      #
      # @example
      #   even_number = context.get_task_field('even_number')
      def get_task_field(field_name)
        @task.context[field_name.to_s]
      end

      # Cross-language standard alias for get_task_field
      # @see #get_task_field
      alias get_input get_task_field

      # Get a field from the task context with a default value.
      #
      # @param field_name [String, Symbol] Field name in task context
      # @param default [Object] Default value if field is nil
      # @return [Object] The field value or default if nil
      #
      # @example
      #   batch_size = context.get_input_or('batch_size', 100)
      def get_input_or(field_name, default = nil)
        value = get_task_field(field_name)
        value.nil? ? default : value
      end

      # Get a configuration value from the handler initialization.
      #
      # @param key [String, Symbol] Configuration key
      # @return [Object, nil] The configuration value or nil if not found
      #
      # @example
      #   api_url = context.get_config('api_url')
      def get_config(key)
        step_config[key.to_s]
      end

      # Get a dependency result from a parent step.
      #
      # This returns the actual result value, not the full metadata hash.
      # Use dependency_results.get_result(step_name) for full metadata.
      #
      # @param step_name [String, Symbol] Name of the parent step
      # @return [Object, nil] The result value or nil if not found
      #
      # @example
      #   prev_result = context.get_dependency_result('step_1')
      def get_dependency_result(step_name)
        @dependency_results.get_results(step_name)
      end

      # Extract a nested field from a dependency result.
      #
      # Useful when dependency results are complex objects and you need
      # to extract a specific nested value.
      #
      # @param step_name [String, Symbol] Name of the parent step
      # @param path [Array<String, Symbol>] Path to the nested field
      # @return [Object, nil] The nested value or nil if not found
      #
      # @example
      #   csv_path = context.get_dependency_field('analyze_csv', 'csv_file_path')
      #   nested_value = context.get_dependency_field('step_1', 'data', 'items', 0)
      def get_dependency_field(step_name, *path)
        result = get_dependency_result(step_name)
        return nil if result.nil?

        result.dig(*path.map(&:to_s))
      end

      # ========================================================================
      # RETRY HELPERS
      # ========================================================================

      # Check if this execution is a retry attempt.
      #
      # @return [Boolean] true if retry_count > 0
      #
      # @example
      #   if context.is_retry?
      #     logger.info("Retrying step, attempt #{context.retry_count}")
      #   end
      def is_retry?
        retry_count.positive?
      end

      # Check if this is the last retry attempt.
      #
      # @return [Boolean] true if this is the final retry attempt
      #
      # @example
      #   if context.is_last_retry?
      #     # Send alert or take special action on final attempt
      #   end
      def is_last_retry?
        retry_count >= max_retries - 1
      end

      # ========================================================================
      # CHECKPOINT ACCESSORS (TAS-125 Batch Processing Support)
      # ========================================================================

      # Get the raw checkpoint data from the workflow step.
      #
      # @return [Hash, nil] The checkpoint data or nil if not set
      def checkpoint
        @workflow_step.checkpoint
      end

      # Get the checkpoint cursor position.
      #
      # The cursor represents the current position in batch processing,
      # allowing handlers to resume from where they left off.
      #
      # @return [Object, nil] The cursor value (int, string, or object)
      #
      # @example
      #   cursor = context.checkpoint_cursor
      #   start_from = cursor || 0
      def checkpoint_cursor
        checkpoint&.dig('cursor')
      end

      # Get the number of items processed in the current batch run.
      #
      # @return [Integer] Number of items processed (0 if no checkpoint)
      def checkpoint_items_processed
        checkpoint&.dig('items_processed') || 0
      end

      # Get the accumulated results from batch processing.
      #
      # Accumulated results allow handlers to maintain running totals
      # or aggregated state across checkpoint boundaries.
      #
      # @return [Hash, nil] The accumulated results or nil if not set
      #
      # @example
      #   totals = context.accumulated_results || {}
      #   current_sum = totals['sum'] || 0
      def accumulated_results
        checkpoint&.dig('accumulated_results')
      end

      # Check if a checkpoint exists for this step.
      #
      # @return [Boolean] true if a checkpoint cursor exists
      #
      # @example
      #   if context.has_checkpoint?
      #     logger.info("Resuming from checkpoint at cursor: #{context.checkpoint_cursor}")
      #   end
      def has_checkpoint?
        !checkpoint_cursor.nil?
      end

      # ========================================================================
      # ADDITIONAL ACCESSORS
      # ========================================================================

      # @return [String, nil] Namespace name from task template
      def namespace_name
        @task.respond_to?(:namespace_name) ? @task.namespace_name : nil
      end

      # @return [String] Step name from workflow step
      def step_name
        @workflow_step.name
      end

      # @return [ActiveSupport::HashWithIndifferentAccess] Full task context
      def context
        @task.context
      end

      # @return [Boolean] Whether the step can be retried
      def retryable?
        @workflow_step.retryable
      end

      # String representation for debugging
      def to_s
        "#<StepContext task_uuid=#{task_uuid} step_uuid=#{step_uuid} step_name=#{step_name}>"
      end

      # Detailed inspection for debugging
      def inspect
        "#<StepContext:#{object_id} " \
          "task_uuid=#{task_uuid.inspect} " \
          "step_uuid=#{step_uuid.inspect} " \
          "step_name=#{step_name.inspect} " \
          "handler_name=#{handler_name.inspect}>"
      end
    end
  end
end

File: workers/typescript/src/types/step-context.ts
import type { FfiStepEvent } from '../ffi/types';

/**
 * Parameters for constructing a StepContext.
 */
export interface StepContextParams {
  event: FfiStepEvent;
  taskUuid: string;
  stepUuid: string;
  correlationId: string;
  handlerName: string;
  inputData: Record<string, unknown>;
  dependencyResults: Record<string, unknown>;
  stepConfig: Record<string, unknown>;
  stepInputs: Record<string, unknown>;
  retryCount: number;
  maxRetries: number;
}

/**
 * Context provided to step handlers during execution.
 *
 * Contains all information needed for a step handler to execute,
 * including input data, dependency results, and configuration.
 *
 * Matches Python's StepContext and Ruby's StepContext (post-TAS-96).
 *
 * @example
 * ```typescript
 * class MyHandler extends StepHandler {
 *   async call(context: StepContext): Promise<StepHandlerResult> {
 *     const orderId = context.inputData['order_id'];
 *     const previousResult = context.getDependencyResult('step_1');
 *     // ... handler logic ...
 *     return this.success({ processed: true });
 *   }
 * }
 * ```
 */
export class StepContext {
  /** The original FFI step event */
  public readonly event: FfiStepEvent;

  /** Task UUID */
  public readonly taskUuid: string;

  /** Step UUID */
  public readonly stepUuid: string;

  /** Correlation ID for tracing */
  public readonly correlationId: string;

  /** Name of the handler being executed */
  public readonly handlerName: string;

  /** Input data for the handler (from task context) */
  public readonly inputData: Record<string, unknown>;

  /** Results from dependent steps */
  public readonly dependencyResults: Record<string, unknown>;

  /** Handler-specific configuration (from step_definition.handler.initialization) */
  public readonly stepConfig: Record<string, unknown>;

  /** Step-specific inputs (from workflow_step.inputs, used for batch cursor config) */
  public readonly stepInputs: Record<string, unknown>;

  /** Current retry attempt number */
  public readonly retryCount: number;

  /** Maximum retry attempts allowed */
  public readonly maxRetries: number;

  constructor(params: StepContextParams) {
    this.event = params.event;
    this.taskUuid = params.taskUuid;
    this.stepUuid = params.stepUuid;
    this.correlationId = params.correlationId;
    this.handlerName = params.handlerName;
    this.inputData = params.inputData;
    this.dependencyResults = params.dependencyResults;
    this.stepConfig = params.stepConfig;
    this.stepInputs = params.stepInputs;
    this.retryCount = params.retryCount;
    this.maxRetries = params.maxRetries;
  }

  /**
   * Create a StepContext from an FFI event.
   *
   * Extracts input data, dependency results, and configuration from
   * the task_sequence_step payload.
   *
   * The FFI data structure mirrors the Ruby TaskSequenceStepWrapper:
   * - task.context -> inputData (task context with user inputs)
   * - dependency_results -> results from parent steps
   * - step_definition.handler.initialization -> stepConfig
   * - workflow_step.attempts -> retryCount
   * - workflow_step.max_attempts -> maxRetries
   * - workflow_step.inputs -> stepInputs
   *
   * @param event - The FFI step event
   * @param handlerName - Name of the handler to execute
   * @returns A StepContext populated from the event
   */
  static fromFfiEvent(event: FfiStepEvent, handlerName: string): StepContext {
    // Extract task context (inputData) from the task structure
    // The task field contains task metadata including context
    const task = event.task ?? {};
    const inputData = (task.context as Record<string, unknown>) ?? {};

    // Extract dependency results
    const dependencyResults = (event.dependency_results as Record<string, unknown>) ?? {};

    // Extract step config from handler initialization
    const stepDefinition = event.step_definition ?? {};
    const handlerConfig = stepDefinition.handler ?? {};
    const stepConfig = (handlerConfig.initialization as Record<string, unknown>) ?? {};

    // Extract retry information and step inputs from workflow_step
    const workflowStep = event.workflow_step ?? {};
    const retryCount = workflowStep.attempts ?? 0;
    const maxRetries = workflowStep.max_attempts ?? 3;
    const stepInputs = (workflowStep.inputs as Record<string, unknown>) ?? {};

    return new StepContext({
      event,
      taskUuid: event.task_uuid,
      stepUuid: event.step_uuid,
      correlationId: event.correlation_id,
      handlerName,
      inputData,
      dependencyResults,
      stepConfig,
      stepInputs,
      retryCount,
      maxRetries,
    });
  }

  /**
   * Get the computed result value from a dependency step.
   *
   * This method extracts the actual computed value from a dependency result,
   * unwrapping any nested structure. Matches Python's get_dependency_result().
   *
   * The dependency result structure can be:
   * - {"result": actual_value} - unwraps to actual_value
   * - primitive value - returns as-is
   *
   * @param stepName - Name of the dependency step
   * @returns The computed result value, or null if not found
   *
   * @example
   * ```typescript
   * // Instead of:
   * const step1Result = context.dependencyResults['step_1'] || {};
   * const value = step1Result.result;  // Might be nested!
   *
   * // Use:
   * const value = context.getDependencyResult('step_1');  // Unwrapped
   * ```
   */
  getDependencyResult(stepName: string): unknown {
    const resultHash = this.dependencyResults[stepName];
    if (resultHash === undefined || resultHash === null) {
      return null;
    }

    // If it's an object with a 'result' key, extract that value
    if (typeof resultHash === 'object' && resultHash !== null && 'result' in resultHash) {
      return (resultHash as Record<string, unknown>).result;
    }

    // Otherwise return the whole thing (might be a primitive value)
    return resultHash;
  }

  /**
   * Get a value from the input data.
   *
   * @param key - The key to look up in inputData
   * @returns The value or undefined if not found
   */
  getInput<T = unknown>(key: string): T | undefined {
    return this.inputData[key] as T | undefined;
  }

  /**
   * Get a value from the step configuration.
   *
   * @param key - The key to look up in stepConfig
   * @returns The value or undefined if not found
   */
  getConfig<T = unknown>(key: string): T | undefined {
    return this.stepConfig[key] as T | undefined;
  }

  /**
   * Check if this is a retry attempt.
   *
   * @returns True if retryCount > 0
   */
  isRetry(): boolean {
    return this.retryCount > 0;
  }

  /**
   * Check if this is the last allowed retry attempt.
   *
   * @returns True if retryCount >= maxRetries - 1
   */
  isLastRetry(): boolean {
    return this.retryCount >= this.maxRetries - 1;
  }

  /**
   * Get a value from the input data with a default.
   *
   * @param key - The key to look up in inputData
   * @param defaultValue - Value to return if key not found or undefined
   * @returns The value or default if not found/undefined
   *
   * @example
   * ```typescript
   * const batchSize = context.getInputOr('batch_size', 100);
   * ```
   */
  getInputOr<T = unknown>(key: string, defaultValue: T): T {
    const value = this.inputData[key];
    return value === undefined ? defaultValue : (value as T);
  }

  /**
   * Extract a nested field from a dependency result.
   *
   * Useful when dependency results are complex objects and you need
   * to extract a specific nested value without manual object traversal.
   *
   * @param stepName - Name of the dependency step
   * @param path - Path elements to traverse into the result
   * @returns The nested value, or null if not found
   *
   * @example
   * ```typescript
   * // Extract nested field from dependency result
   * const csvPath = context.getDependencyField('analyze_csv', 'csv_file_path');
   * // Multiple levels deep
   * const value = context.getDependencyField('step_1', 'data', 'items');
   * ```
   */
  getDependencyField(stepName: string, ...path: string[]): unknown {
    let result = this.getDependencyResult(stepName);
    if (result === null || result === undefined) {
      return null;
    }
    for (const key of path) {
      if (typeof result !== 'object' || result === null) {
        return null;
      }
      result = (result as Record<string, unknown>)[key];
    }
    return result;
  }

  // ===========================================================================
  // CHECKPOINT ACCESSORS (TAS-125 Batch Processing Support)
  // ===========================================================================

  /**
   * Get the raw checkpoint data from the workflow step.
   *
   * @returns The checkpoint data object or null if not set
   */
  get checkpoint(): Record<string, unknown> | null {
    const workflowStep = this.event.workflow_step ?? {};
    return (workflowStep.checkpoint as Record<string, unknown>) ?? null;
  }

  /**
   * Get the checkpoint cursor position.
   *
   * The cursor represents the current position in batch processing,
   * allowing handlers to resume from where they left off.
   *
   * @returns The cursor value (number, string, or object) or null if not set
   *
   * @example
   * ```typescript
   * const cursor = context.checkpointCursor;
   * const startFrom = cursor ?? 0;
   * ```
   */
  get checkpointCursor(): unknown {
    return this.checkpoint?.cursor ?? null;
  }

  /**
   * Get the number of items processed in the current batch run.
   *
   * @returns Number of items processed (0 if no checkpoint)
   */
  get checkpointItemsProcessed(): number {
    return (this.checkpoint?.items_processed as number) ?? 0;
  }

  /**
   * Get the accumulated results from batch processing.
   *
   * Accumulated results allow handlers to maintain running totals
   * or aggregated state across checkpoint boundaries.
   *
   * @returns The accumulated results object or null if not set
   *
   * @example
   * ```typescript
   * const totals = context.accumulatedResults ?? {};
   * const currentSum = totals.sum ?? 0;
   * ```
   */
  get accumulatedResults(): Record<string, unknown> | null {
    return (this.checkpoint?.accumulated_results as Record<string, unknown>) ?? null;
  }

  /**
   * Check if a checkpoint exists for this step.
   *
   * @returns True if a checkpoint cursor exists
   *
   * @example
   * ```typescript
   * if (context.hasCheckpoint()) {
   *   console.log(`Resuming from cursor: ${context.checkpointCursor}`);
   * }
   * ```
   */
  hasCheckpoint(): boolean {
    return this.checkpointCursor !== null;
  }

  /**
   * Get all dependency result keys.
   *
   * @returns Array of step names that have dependency results
   */
  getDependencyResultKeys(): string[] {
    return Object.keys(this.dependencyResults);
  }

  /**
   * Get all dependency results matching a step name prefix.
   *
   * This is useful for batch processing where multiple worker steps
   * share a common prefix (e.g., "process_batch_001", "process_batch_002").
   *
   * Returns the unwrapped result values (same as getDependencyResult).
   *
   * @param prefix - Step name prefix to match
   * @returns Array of unwrapped result values from matching steps
   *
   * @example
   * ```typescript
   * // For batch worker results named: process_batch_001, process_batch_002, etc.
   * const batchResults = context.getAllDependencyResults('process_batch_');
   * const total = batchResults.reduce((sum, r) => sum + r.count, 0);
   * ```
   */
  getAllDependencyResults(prefix: string): unknown[] {
    const results: unknown[] = [];

    for (const key of Object.keys(this.dependencyResults)) {
      if (key.startsWith(prefix)) {
        const result = this.getDependencyResult(key);
        if (result !== null) {
          results.push(result);
        }
      }
    }

    return results;
  }
}
